libnvme/src/firmware.rs

```diff
@@ -0,0 +1,306 @@
+            }
+            let cstring = unsafe { CString::from_vec_unchecked(bytes) };
+            // NVMe Spec: "The firmware revision is indicated as an ASCII
+            // string."
+            let firmware = cstring.to_string_lossy().trim().to_string();
+            firmware_slots.push(Some(firmware));
+        }
+
+        Self {
+            active_slot,
+            next_active_slot,
+            slot1_is_read_only,
+            number_of_slots,
+            firmware_slots,
+        }
+    }
+
+    /// Get the firmware version for a particular slot.
+    ///
+    /// Note that the NVMe spec allows for slots 1 - 7.
+    pub fn get_slot<S>(&self, slot: S) -> Result<Option<&String>, NvmeSlotError>
+    where
+        S: TryInto<NvmeSlot>,
+        S::Error: Into<NvmeSlotError>,
+    {
+        let slot = slot.try_into().map_err(Into::into)?;
+        // We subtract 1 because our internal mapping is 0 indexed.
+        match self.firmware_slots.get(slot.0 as usize - 1) {
+            Some(slot) => Ok(slot.as_ref()),
+            None => Err(NvmeSlotError::DoesNotExisit(slot.0)),
+        }
+    }
+
+    /// Returns an iterator over the firmware slots.
+    ///
+    /// The iterator yields a `Some(String)` if the slot has a firmware version
+    /// commited otherwise it yeilds `None`.
+    pub fn slot_iter(&self) -> impl Iterator<Item = Option<&String>> {
+        self.firmware_slots.iter().map(|o| o.as_ref())
+    }
+}
+
+impl<'a> Controller<'a> {
+    /// Get the controller's firmware log page.
+    pub fn get_firmware_log_page(&self) -> Result<FirmwareLogPage, NvmeError> {
+        let LogPageInfo { size, req, .. } =
+            self.get_logpage(LogPageName::Firmware)?;
+        // TODO the size of nvme_fwslot_log_t should always match the size
+        // returned from get_logpage. We may want a custom error type here if
+        // that's not true. Although for now libnvme should error if we give it
+        // something too small.
+        let mut buf = MaybeUninit::<nvme_fwslot_log_t>::zeroed();
+        self.check_result(
+            unsafe {
+                nvme_log_req_set_output(req.inner, buf.as_mut_ptr() as _, size)
+            },
+            || format!("failed to set logpage req size to {size}"),
+        )?;
+
+        self.check_result(unsafe { nvme_log_req_exec(req.inner) }, || {
+            "failed to execute firmware log request"
+        })?;
+
+        // Our request was successful so we can assume this struct has been
+        // populated now.
+        let logpage = unsafe { buf.assume_init() };
+        let controller_info = self.get_info()?;
+        let identify = controller_info.get_controller_info_identify();
+        Ok(FirmwareLogPage::init(&identify, logpage))
+    }
+}
+
+impl<'ctrl> WriteLockedController<'ctrl> {
+    /// Upload new firmware to the NVMe controller.
+    ///
+    /// Note this firmware needs to be commited to a slot via
+    /// `FirmwareCommitRequestBuilder`.
```

Comment:
I don't entirely follow this note - can you explain? I'm wondering if the data loading should be rolled into `FirmwareCommitRequestBuilder`.

Reply:
Sure! 

The NVMe spec defines these as separate admin actions.  There's a firmware load operation, a firmware commit operation, and finally a firmware activation operation.  An operator first uploads the firmware blob to the device (where it goes on the device is totally up to the manufacture), and then the operator must commit the firmware to one of the slots.

In `nvmeadm` that looks like this:
```
  load-firmware       - load firmware to a controller
  commit-firmware     - commit downloaded firmware to a slot of a controller
  activate-firmware   - activate a firmware slot of a controller
```

In this libnvme crate that looks like this:
```rust
        // Upload and activate the firmware
        controller.firmware_load(fw_bytes)?;

        // Set slot and activate (combined steps)
        controller
            .firmware_commit_request()?
            .set_slot(NvmeSlot::try_from(next_slot)?)?
            .set_action(libnvme::firmware::FirmwareCommitAction::SaveActivate)? // <--- can be separate reqs 
            .execute()?;
```